# Business Understanding

## Overview

# Data Understanding

The data being used is from a Credit Bereau. and it consists of:

**Application Data:**
* **Application_ID**: Unique Application Number
* **Application_Date**: Date of application
* **Loan_Approval_Status**: Approval status of loan
* **Age_at_Application**: Age in Years at point of application
* **Marital_Status**: Marital status at point of application
* **Gender**: Sex
* **Own_Rent_Home**: Residential status of home ownership
* **Occupation_Professional_Category**: Proffessional Category
* **Type_of_Business_Industry_of_Employment**: Industry of Employment
* **Employment_Status**: Employment status
* **Time_at_Current_Employment_(Months)**: Time at current employment in months
* **Time_Employed_with_Previous_Employer_(Months)**: Time at previous employment in months
* **Number_of_Dependents**: Number of dependents
* **Loan_Purpose**: Purpose of Loan Applied
* **Monthly_Income**: Monthly Income in KES
* **Government_Employee**: Flag if government employee

**Contracts Data:**
* **Contract_ID**:	Loan Account Number
* **Application_ID**	Application Number related to the loan opened
* **Key**:	Number of Open Contracts related to the loan application Number
* **Loan_Type**:	Type of loan facility
* **Contract_Start_Date**:	Date of start for Loan facility
* **Expected_Contract_End_Date**:	Expected date of loan maturity
* **Actual_Contract_End_Date**:	Actual date when the contrcat was closed
* **Loan_Term**:	Tenure of loan in months
* **Total_Loan_Amount**:	Loan amount approved

**Contracts Snapshop data:**
* **Contract Id**:	Loan Account Number*
* **Snapshot Date**:	The End of Month when the data was reported
* **Outstanding Amount**:	The outstanding balance as at the reporting date
* **Due Amount**:	The amount unpaid as at the reporting date
* **Number of Days Past Due**:	The number of days past due as at that reporting date
* **Contract Status**:	The status of the contract

**CRB Data:**
* **Application_ID**:	Application Number related to the loan opened
* **Application_Date**:	Date of application
* **CRB Score**:	Credit Bureau Risk Score
* **CRB Grade**:	Credit Bureau Risk Grade

**Current and Savings Account Data:**
* **Account Number**:	Application Number related to the loan opened
* **Application ID**:	Application Number related to the loan opened
* **Month (YYYY-MM)**:	The reporting month when the data was calculated
* **Open Date**:	The date when the account was openned
* **Closed Date**:	The date when the account was closed
* **Average Balance**:	The monthly average balance
* **Overdraft Limit**:	Overdraft Limit
* **Number of Days in Debit during Month**:	Number of Days in Debit during Month
* **Number of Days in Credit during Month**: Number of Days in Credit during Month
* **Maximum End of Day Balance**:	Maximum End of Day Balance
* **Minimum End of Day Balance**:	Minimum End of Day Balance
* **Sum of Monthly Debit Transactions**: Sum of Monthly Debit Transactions
* **Sum of Monthly Credit Transactions**:	Sum of Monthly Credit Transactions
* **Last Transaction Date**:	Last Transaction Date
* **Sum of ATM Deposits**:	Sum of ATM Deposits
* **Number of ATM Deposits**:	Number of ATM Deposits
* **Sum of ATM Withdrawals**:	Sum of ATM Withdrawals
* **Number of ATM Withdrawals**:	Number of ATM Withdrawals
* **Sum of Cheque Debit Transactions**:	Sum of Cheque Debit Transactions
* **Number of Cheque Debit Transactions**:	Number of Cheque Debit Transactions
* **Sum of Bounced Cheque transactions**:	Sum of Bounced Cheque transactions
* **Number of Bounced Cheques**:	Number of Bounced Cheques
* **Sum of Debit Card POS Transactions**:	Sum of Debit Card POS Transactions
* **Number of Debit Card POS Transactions**:	Number of Debit Card POS Transactions
* **Sum of Over the Counter Withdrawals**:	Sum of Over the Counter Withdrawals
* **Number of Over the Counter withdrawals**:	Number of Over the Counter withdrawals
* **Account Type**:	Type of account
* **Account Status**:	Latest status of the account
* **Number of Joint Holders**:	Number of Persons owning the account
* **SnapshotMonth**:	The reporting month when the data was calculated

![image.png](..\images\erpdiagram.png)

In [67]:
# Loading the relevant libraries

# Data pre-processsing libraries
import pandas as pd
import numpy as np
import re

# Metrics libraries

# Data modeling libraries

# 

### Application Data

In [68]:
# Load data
app_data = pd.read_csv("./data/1. ApplicationData.csv")

# Check 2 top rows
app_data.head(2)

Application_ID Application_Date Loan_Approval_Status  Age_at_Application  \
0     C302100001       2015-09-15             Approved                39.0   
1     C302100002       2015-08-31             Approved                29.0   

  Marital_Status  Gender Own_Rent_Home Occupation_Professional_Category  \
0      Unmarried  Female           Own                              NaN   
1      Unmarried  Female        Family                              NaN   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                                   Other         Full-Time   
1                                     NaN         Full-Time   

   Time_at_Current_Employment_(Months)  \
0                                  6.0   
1                                 25.0   

   Time_Employed_with_Previous_Employer_(Months)  Number_of_Dependents  \
0                                            NaN                   2.0   
1                                            NaN                   0.0   

         Loan_Purpose  Monthly_Income Government_Employee  
0    Home Improvement         1500.59                 NaN  
1  Debt Consolidation         3769.36                 NaN

In [69]:
# Get data summary
app_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9492 entries, 0 to 9491
Data columns (total 16 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Application_ID                                 9492 non-null   object 
 1   Application_Date                               9492 non-null   object 
 2   Loan_Approval_Status                           9354 non-null   object 
 3   Age_at_Application                             9471 non-null   float64
 4   Marital_Status                                 7712 non-null   object 
 5   Gender                                         9488 non-null   object 
 6   Own_Rent_Home                                  8188 non-null   object 
 7   Occupation_Professional_Category               9361 non-null   object 
 8   Type_of_Business_Industry_of_Employment        7210 non-null   object 
 9   Employment_Status                              9357 

In [70]:
# Check Shape
app_data.shape

(9492, 16)

In [71]:
# Check null values
app_data.isnull().sum()

Application_ID                                      0
Application_Date                                    0
Loan_Approval_Status                              138
Age_at_Application                                 21
Marital_Status                                   1780
Gender                                              4
Own_Rent_Home                                    1304
Occupation_Professional_Category                  131
Type_of_Business_Industry_of_Employment          2282
Employment_Status                                 135
Time_at_Current_Employment_(Months)               232
Time_Employed_with_Previous_Employer_(Months)    9194
Number_of_Dependents                             1814
Loan_Purpose                                       42
Monthly_Income                                      0
Government_Employee                               181
dtype: int64

*Observation*

* The ApplicationData has $9,492$ records under $16$ columns.
* All the columns have null values apart from the *'Application_ID'*, *'Application_Date'*, and *'Monthly_Income'* columns.
* The 'Application_Date' column is  an object dtype. thiss needs to be changed to 'date' datatype

### Contracts Data

In [72]:
# Load data
contract_data = pd.read_csv("./data/2. ContractsData.csv")

# Check 2 top rows
contract_data.head(2)

Contract_ID Application_ID  Key                Loan_Type  \
0  CONT100001     C302103426    1  Unsecured Personal Loan   
1  CONT100002     C302103426    2  Unsecured Personal Loan   

  Contract_Start_Date Expected_Contract_End_Date Actual_Contract_End_Date  \
0          2018-04-12                 2023-04-12                      NaN   
1          2018-04-12                 2023-04-12                      NaN   

       Loan_Term  Total_Loan_Amount  
0  60 - Month(s)           24529.89  
1  60 - Month(s)           35085.79

In [73]:
# show data summary
contract_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract_ID                 9374 non-null   object 
 1   Application_ID              7598 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Loan_Type                   9374 non-null   object 
 4   Contract_Start_Date         9374 non-null   object 
 5   Expected_Contract_End_Date  9374 non-null   object 
 6   Actual_Contract_End_Date    2593 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


In [74]:
# Check shape
contract_data.shape

(9374, 9)

In [75]:
# Check null values
contract_data.isnull().sum()

Contract_ID                      0
Application_ID                1776
Key                              0
Loan_Type                        0
Contract_Start_Date              0
Expected_Contract_End_Date       0
Actual_Contract_End_Date      6781
Loan_Term                        0
Total_Loan_Amount                0
dtype: int64

*Observation*
* The contract data is of $9$ columns with $9,374$ records
* The dates are in the object dtype, will need to be changed to date format
* The contracts data has null values on the *'Application_ID'* and *'Actual_Contract_End_Date'* columns.


### Contracts Snapshot data

In [76]:
# Load the dataset
contractsSnapshot_data = pd.read_csv("./data/3. ContractsSnapshotData.csv")

# Check top 2 rows
contractsSnapshot_data.head(2)

Contract_ID Snapshot Date  Outstanding Amount  Due Amount  \
0  CONT109375    2018-12-31                0.00         0.0   
1  CONT109376    2006-08-31              215.08         0.0   

   Number of Days Past Due Contract Status  
0                      NaN          Closed  
1                      NaN          Active

In [77]:
# Check Data size
contractsSnapshot_data.shape

(274371, 6)

In [78]:
# Data summary
contractsSnapshot_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274371 entries, 0 to 274370
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Contract_ID              274371 non-null  object 
 1   Snapshot Date            274371 non-null  object 
 2   Outstanding Amount       274371 non-null  float64
 3   Due Amount               274371 non-null  float64
 4   Number of Days Past Due  79158 non-null   float64
 5   Contract Status          274371 non-null  object 
dtypes: float64(3), object(3)
memory usage: 12.6+ MB


In [79]:
# Check null values
contractsSnapshot_data.isnull().sum()

Contract_ID                     0
Snapshot Date                   0
Outstanding Amount              0
Due Amount                      0
Number of Days Past Due    195213
Contract Status                 0
dtype: int64

*Observation*
* Tha contractsSnapshot data has $6$ columns with $274,371$ rows.
* the date column is of object dtype. Will be changed to date dtype
* It has null values on the *'Number of Days Past Due'* column. On further checking, the null values are present on records that do not have *'Due Amount'*.

### CRB Data

In [80]:
# Load data
crb_data = pd.read_csv("./data/4. CRB Data.csv")

# Check top 2 rows
crb_data.head(2)

Application_ID Application_Date  CRB Score CRB Grade
0     C302102137       2017-09-30        593         D
1     C302106175       2017-09-30        625         C

In [81]:
# Data summary
crb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6750 entries, 0 to 6749
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Application_ID    6750 non-null   object
 1   Application_Date  6750 non-null   object
 2   CRB Score         6750 non-null   int64 
 3   CRB Grade         6750 non-null   object
dtypes: int64(1), object(3)
memory usage: 211.1+ KB


In [82]:
# Check null values
crb_data.isnull().sum()

Application_ID      0
Application_Date    0
CRB Score           0
CRB Grade           0
dtype: int64

In [83]:
# Check size
crb_data.shape

(6750, 4)

*Observation*
* CRB Data has $4$ columns with $6,750$ records.
* The CRB Data has no null values.
* The *'Application_Date'* column is of object dtype, to be changed to date dtype.

### Current_and_Savings Data

In [84]:
# load data
curr_savings_data = pd.read_csv("./data/5. Current and Savings Account Data.csv")

# Check top 2 rows
curr_savings_data.head(2)

ID Account_Number Application ID Month (YYYY-MM)   Open Date Closed Date  \
0   1    ACC10000001     C302100001      2015-09-01  2015-09-06         NaN   
1   2    ACC10000002     C302100001      2015-09-01  2015-09-06  2018-10-04   

   Average Balance  Overdraft Limit  Number of Days in Debit during Month  \
0          1986.09              0.0                                   0.0   
1            23.45              NaN                                   0.0   

   Number of Days in Credit during Month  ...  \
0                                   25.0  ...   
1                                   25.0  ...   

   Sum of Bounced Cheque transactions  Number of Bounced Cheques  \
0                                 NaN                        0.0   
1                                 NaN                        0.0   

   Sum of Debit Card POS Transactions  Number of Debit Card POS Transactions  \
0                                 NaN                                    0.0   
1                                 NaN                                    0.0   

  Sum of Over the Counter Withdrawals  Number of Over the Counter withdrawals  \
0                                 NaN                                     0.0   
1                                 NaN                                     0.0   

   Account Type  Account Status  Number of Joint Holders  SpshotMonth  
0       Savings          Active                      0.0   2015-09-01  
1       Savings          Closed                      0.0   2015-09-01  

[2 rows x 31 columns]

In [85]:
# Dataset summary
curr_savings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 31 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      857677 non-null  int64  
 1   Account_Number                          857677 non-null  object 
 2   Application ID                          857672 non-null  object 
 3   Month (YYYY-MM)                         857647 non-null  object 
 4   Open Date                               857647 non-null  object 
 5   Closed Date                             161748 non-null  object 
 6   Average Balance                         857647 non-null  float64
 7   Overdraft Limit                         753484 non-null  float64
 8   Number of Days in Debit during Month    857647 non-null  float64
 9   Number of Days in Credit during Month   857647 non-null  float64
 10  Maximum End of Day Balance              8576

In [86]:
# Check size
curr_savings_data.shape

(857677, 31)

In [87]:
# Check null values
# ~
curr_savings_data.isnull().sum().sort_values()

ID                                             0
Account_Number                                 0
Application ID                                 5
Account Status                                30
Account Type                                  30
Number of Over the Counter withdrawals        30
Number of Debit Card POS Transactions         30
Number of Bounced Cheques                     30
Number of Cheque Debit Transactions           30
Number of ATM Withdrawals                     30
Number of ATM Deposits                        30
Number of Joint Holders                       30
SpshotMonth                                   30
Minimum End of Day Balance                    30
Maximum End of Day Balance                    30
Number of Days in Credit during Month         30
Month (YYYY-MM)                               30
Number of Days in Debit during Month          30
Average Balance                               30
Open Date                                     30
Sum of Monthly Credi

*Observation*
* The Current & Savings data has $31$ columns with $857,677$ records
* This data has null values on all the columns apart from the *'ID'* and *'Account_Number'* columns
* The date columns are also of object dtype. To be changed to date dtype

## Important features

* **ApplicationData.csv**: Loan_Approval_Status, Monthly_Income, Number_of_Dependents, Time_at_Current_Employment_(Months), Employment_Status
* **ContractsData.csv**: Loan_Type, Loan_Term, Total_Loan_Amount
* **ContractsSnapshotData.csv**: Outstanding Amount, Number of Days Past Due, Contract Status
* **CRB Data.csv**: CRB Score, CRB Grade
* **Current and Savings Account Data.csv**: Average Balance, Overdraft Limit, Sum of Monthly Credit Transactions, Number of Bounced Cheques


## Other

In [88]:
app_data.columns

Index(['Application_ID', 'Application_Date', 'Loan_Approval_Status',
       'Age_at_Application', 'Marital_Status', 'Gender', 'Own_Rent_Home',
       'Occupation_Professional_Category',
       'Type_of_Business_Industry_of_Employment', 'Employment_Status',
       'Time_at_Current_Employment_(Months)',
       'Time_Employed_with_Previous_Employer_(Months)', 'Number_of_Dependents',
       'Loan_Purpose', 'Monthly_Income', 'Government_Employee'],
      dtype='object')

In [89]:
rel_app_data = app_data[["Loan_Approval_Status", "Monthly_Income", "Number_of_Dependents", "Time_at_Current_Employment_(Months)", "Employment_Status"]]

# Data Cleaning / Preparation

Understand the data
1. Drop irrelevant columns
2. Handle duplicates
3. Handle missing data
4. Convert data types (e.g., datetime, categorical)
5. Normalize/encode data
6. Merge datasets (if applicable)
7. Feature engineering

1. Handle missing data
2. Convert data types (e.g., datetime, categorical)

In [90]:
# Function to get numerical columns
def get_numerical_columns(df):
    return df.select_dtypes(include=['number']).columns.tolist()

# Function to get categorical columns
def get_categorical_columns(df):
    return df.select_dtypes(include=['object', 'category']).columns.tolist()

def calc_missing_values_percentage(df):
    
    # Calculate % of missing values
    missing_percentage = df.isnull().sum() * 100 / len(df)
    
    # Filter only columns with  50% missing values
    missing_percentage = missing_percentage[missing_percentage > 50] 
    
    # Sort in descending order
    missing_percentage = missing_percentage.sort_values(ascending=False)  

    return pd.DataFrame({'Column': missing_percentage.index, 'Missing_Percentage': missing_percentage.values})

# Create function to fix dates
def fix_date_format(date_str):
    # Check 4th character, if '-' remove it, otherwise return original string
    return re.sub(r'^(...)-', r'\1', date_str)

def change_to_date(df, date_cols):
    
    # Loop through list to change column dtype to datetime
    for col in date_cols:
        df[col] = pd.to_datetime(df[col]).astype("datetime64[D]")

# Function to replace leap year dates
def fix_leap_year_dates(df, columns):
    for col in columns:
        df.loc[:, col] = df[col].str.replace(r"(\d{4})-02-29", r"\1-02-28", regex=True)
        


### Handle missing data

#### Application data - missing values

In [91]:
# Check null values
app_data.isnull().sum()

Application_ID                                      0
Application_Date                                    0
Loan_Approval_Status                              138
Age_at_Application                                 21
Marital_Status                                   1780
Gender                                              4
Own_Rent_Home                                    1304
Occupation_Professional_Category                  131
Type_of_Business_Industry_of_Employment          2282
Employment_Status                                 135
Time_at_Current_Employment_(Months)               232
Time_Employed_with_Previous_Employer_(Months)    9194
Number_of_Dependents                             1814
Loan_Purpose                                       42
Monthly_Income                                      0
Government_Employee                               181
dtype: int64

In [92]:
# Get percentage of missing values above 50%
missing_app_data_values = calc_missing_values_percentage(app_data)
print(missing_app_data_values)

                                          Column  Missing_Percentage
0  Time_Employed_with_Previous_Employer_(Months)           96.860514


In [93]:
# Drop the column with 50% and above missing data
app_data.drop(columns='Time_Employed_with_Previous_Employer_(Months)', inplace=True)

In [94]:
# get numerical columns
numerical_cols = get_numerical_columns(app_data)
# print(numerical_cols)

# Replace all numerical columns with median
for col in numerical_cols:
    app_data[col].fillna(app_data[col].median(), inplace=True)

We will fill the null categorical values with 'unknown' apart from 'Gender' that we will add 'Other' as a category

In [95]:
# Get categorical columns without the primary keys - Application_ID
categorical_cols = ['Loan_Approval_Status', 'Marital_Status', 'Own_Rent_Home', 
                    'Occupation_Professional_Category',
                    'Type_of_Business_Industry_of_Employment', 'Employment_Status', 
                    'Loan_Purpose', 'Government_Employee']

for cat_col in categorical_cols:
    # app_data[cat_col].fillna(app_data["Application ID"].mode()[0], inplace=True)
    app_data[cat_col].fillna("unknown", inplace=True)

In [96]:
# Check unique values for Gender
app_data['Gender'].value_counts()

Male      4845
Female    4643
Name: Gender, dtype: int64

In [97]:
# Replace Null values with Other
app_data["Gender"].fillna("Other", inplace=True)

In [98]:
# Confirm
app_data.isnull().sum()

Application_ID                             0
Application_Date                           0
Loan_Approval_Status                       0
Age_at_Application                         0
Marital_Status                             0
Gender                                     0
Own_Rent_Home                              0
Occupation_Professional_Category           0
Type_of_Business_Industry_of_Employment    0
Employment_Status                          0
Time_at_Current_Employment_(Months)        0
Number_of_Dependents                       0
Loan_Purpose                               0
Monthly_Income                             0
Government_Employee                        0
dtype: int64

#### Contract data - missing values

In [99]:
# Check null values
contract_data.isnull().sum()

Contract_ID                      0
Application_ID                1776
Key                              0
Loan_Type                        0
Contract_Start_Date              0
Expected_Contract_End_Date       0
Actual_Contract_End_Date      6781
Loan_Term                        0
Total_Loan_Amount                0
dtype: int64

The null values are in the "Application_ID" and "Actual_Contract_End_Date".

Since end date is not indicated, we can replace with latest snapshopt date.

In [100]:
# Get max snapshot date per contract
latest_snapshot = contractsSnapshot_data.groupby("Contract_ID")["Snapshot Date"].transform('max')
latest_snapshot

# Fill missing 'Actual_End_Date' with the latest snapshot date per 'Contract_ID'
contract_data['Actual_Contract_End_Date'].fillna(latest_snapshot, inplace=True)

# Confirm
contract_data.isnull().sum()

Contract_ID                      0
Application_ID                1776
Key                              0
Loan_Type                        0
Contract_Start_Date              0
Expected_Contract_End_Date       0
Actual_Contract_End_Date         0
Loan_Term                        0
Total_Loan_Amount                0
dtype: int64

Since we cannot recreate or auto-geneate the primary key, then we can drop the records missing 'Application_ID'. 

In [101]:
# Drop rows with no Application_ID
contract_data.dropna(subset=["Application_ID"], inplace=True)

# Confirm
contract_data.isnull().sum()

Contract_ID                   0
Application_ID                0
Key                           0
Loan_Type                     0
Contract_Start_Date           0
Expected_Contract_End_Date    0
Actual_Contract_End_Date      0
Loan_Term                     0
Total_Loan_Amount             0
dtype: int64

#### ContractsSnapshotData - Missing Values

In [102]:
calc_missing_values_percentage(contractsSnapshot_data)

Column  Missing_Percentage
0  Number of Days Past Due           71.149283

In [103]:
# Check missing values
contractsSnapshot_data.isnull().sum()

Contract_ID                     0
Snapshot Date                   0
Outstanding Amount              0
Due Amount                      0
Number of Days Past Due    195213
Contract Status                 0
dtype: int64

Checking the records missing number of days due,most of them have $0$ Due amount. So let us replace this with a $0$

In [104]:
# Check other features where "Number of Days Past Due" is null
contractsSnapshot_data[contractsSnapshot_data["Number of Days Past Due"].isnull()]

Contract_ID Snapshot Date  Outstanding Amount  Due Amount  \
0       CONT109375    2018-12-31                0.00         0.0   
1       CONT109376    2006-08-31              215.08         0.0   
2       CONT109376    2006-09-30              215.08         0.0   
3       CONT109376    2006-10-31              215.08         0.0   
4       CONT109376    2006-11-30              215.08         0.0   
...            ...           ...                 ...         ...   
274366  CONT116311   202-1-08-31           117233.29         0.0   
274367  CONT116312   202-1-08-31            43789.21         0.0   
274368  CONT116313   202-1-08-31           160023.45         0.0   
274369  CONT116314   202-1-08-31           140679.95         0.0   
274370  CONT116315   202-1-08-31            70339.98         0.0   

        Number of Days Past Due Contract Status  
0                           NaN          Closed  
1                           NaN          Active  
2                           NaN          Active  
3                           NaN          Active  
4                           NaN          Active  
...                         ...             ...  
274366                      NaN          Active  
274367                      NaN          Active  
274368                      NaN          Active  
274369                      NaN          Active  
274370                      NaN          Active  

[195213 rows x 6 columns]

In [105]:
# Fill in with a '0'
contractsSnapshot_data['Number of Days Past Due'].fillna(0, inplace=True)

# Confirm
contractsSnapshot_data.isnull().sum()

Contract_ID                0
Snapshot Date              0
Outstanding Amount         0
Due Amount                 0
Number of Days Past Due    0
Contract Status            0
dtype: int64

CRB Data has no null values. So let us work on 'Current & Savings' data.

#### Current and Savings Data - missing values

In [106]:
# Check % of missing data
curr_savings_cols_to_drop = calc_missing_values_percentage(curr_savings_data)
curr_savings_cols_to_drop

Column  Missing_Percentage
0   Sum of Bounced Cheque transactions           99.999767
1     Sum of Cheque Debit Transactions           99.926079
2                  Sum of ATM Deposits           98.294929
3  Sum of Over the Counter Withdrawals           92.530521
4   Sum of Debit Card POS Transactions           86.628766
5                          Closed Date           81.141152
6               Sum of ATM Withdrawals           77.997661

In [107]:
# Drop Columns with more than 50% missing values

# get list of columns
curr_savings_cols_to_drop = curr_savings_cols_to_drop['Column']

# Drop columns
curr_savings_data.drop(columns=curr_savings_cols_to_drop, inplace=True)

In [108]:
curr_savings_data.isnull().sum()

ID                                             0
Account_Number                                 0
Application ID                                 5
Month (YYYY-MM)                               30
Open Date                                     30
Average Balance                               30
Overdraft Limit                           104193
Number of Days in Debit during Month          30
Number of Days in Credit during Month         30
Maximum End of Day Balance                    30
Minimum End of Day Balance                    30
Sum of Monthly Debit Transactions          24686
Sum of Monthly Credit Transactions           148
Last Transaction Date                     332006
Number of ATM Deposits                        30
Number of ATM Withdrawals                     30
Number of Cheque Debit Transactions           30
Number of Bounced Cheques                     30
Number of Debit Card POS Transactions         30
Number of Over the Counter withdrawals        30
Account Type        

In [109]:
# Create a list of numerical columns without ID/index columns
curr_savings_numerical_cols = get_numerical_columns(curr_savings_data)
curr_savings_numerical_cols = ['Average Balance', 'Overdraft Limit', 
 'Number of Days in Debit during Month', 'Number of Days in Credit during Month',
 'Maximum End of Day Balance', 'Minimum End of Day Balance',
 'Sum of Monthly Debit Transactions', 'Sum of Monthly Credit Transactions',
 'Number of ATM Deposits', 'Number of ATM Withdrawals',
 'Number of Cheque Debit Transactions', 'Number of Bounced Cheques',
 'Number of Debit Card POS Transactions', 'Number of Over the Counter withdrawals',
 'Number of Joint Holders']

# Create a list of categorical columns without ID/index columns
curr_savings_cat_cols =get_categorical_columns(curr_savings_data)
curr_savings_cat_cols = ['Month (YYYY-MM)', 'Account Type', 'Account Status', 'SpshotMonth']

Since most caegorical columns values are unknown, we can fill in with 'unknown'

Let us fill in numerical columns with '0'

In [110]:
# Fill missing values
for cat_col in curr_savings_cat_cols:
    curr_savings_data[cat_col].fillna("unknown", inplace=True)
    
for num_col in curr_savings_numerical_cols:
    curr_savings_data[num_col].fillna(0, inplace=True)

In [111]:
# Drop columns with no primary key - Application ID
curr_savings_data.dropna(subset=["Application ID"], inplace=True)


In [112]:
# Replace dates with 
# Create a list of date columns
curr_savings_date_cols = ["Open Date", "Last Transaction Date"]

# Fill missing dates
for col in curr_savings_date_cols:
    curr_savings_data[col].fillna("1900-01-01", inplace=True)
    # curr_savings_data["Closed Date"].fillna("1900-01-01", inplace=True)

# Confirm
curr_savings_data.isna().sum()

ID                                        0
Account_Number                            0
Application ID                            0
Month (YYYY-MM)                           0
Open Date                                 0
Average Balance                           0
Overdraft Limit                           0
Number of Days in Debit during Month      0
Number of Days in Credit during Month     0
Maximum End of Day Balance                0
Minimum End of Day Balance                0
Sum of Monthly Debit Transactions         0
Sum of Monthly Credit Transactions        0
Last Transaction Date                     0
Number of ATM Deposits                    0
Number of ATM Withdrawals                 0
Number of Cheque Debit Transactions       0
Number of Bounced Cheques                 0
Number of Debit Card POS Transactions     0
Number of Over the Counter withdrawals    0
Account Type                              0
Account Status                            0
Number of Joint Holders         

### Convert date data to right format

From the observation, all date data types are in text format. Let us change all of them to the right format.

But before that, you'll notice some dates are not correctly recorded e.g. '*202-1-03-06*' which I think should be '*2021-03-06*'. So we will format it too.

From another observation, leap years were not recognized, so will adjust the dates to fit leap dates as it will not convert to datetime type if it is out of range.

In [113]:
# an example
app_data[9446:9450]

Application_ID Application_Date   Loan_Approval_Status  \
9446     C302109447      202-1-03-06                unknown   
9447     C302109448       2018-07-05                Decline   
9448     C302109449      202-1-02-21  Customer Cancellation   
9449     C302109450       2018-08-22  Customer Cancellation   

      Age_at_Application Marital_Status  Gender Own_Rent_Home  \
9446                22.0        unknown  Female       unknown   
9447                31.0        unknown    Male       unknown   
9448                24.0        unknown    Male       unknown   
9449                37.0        unknown  Female       unknown   

     Occupation_Professional_Category Type_of_Business_Industry_of_Employment  \
9446            Elementary Occupation                                 unknown   
9447                          Officer                                 unknown   
9448                          Officer                                 unknown   
9449                          Officer                                 unknown   

     Employment_Status  Time_at_Current_Employment_(Months)  \
9446          Contract                                 55.0   
9447         Full-Time                                115.0   
9448         Full-Time                                 41.0   
9449         Full-Time                                134.0   

      Number_of_Dependents Loan_Purpose  Monthly_Income Government_Employee  
9446                   1.0        Other         1463.56               False  
9447                   1.0        Other         3657.43                True  
9448                   1.0        Other         3296.86               False  
9449                   1.0        Other         3749.84                True

In [114]:
# # Leap year Example
# contract_data[contract_data["Expected_Contract_End_Date"]=="2021-02-29"]

#### Application data dates format

In [115]:
# list of date_cols 
app_data_date_cols = ['Application_Date']

# Apply the function to fix the date column
app_data['Application_Date'] = app_data['Application_Date'].apply(fix_date_format)

#Call function to change to datetime format
change_to_date(app_data, app_data_date_cols)


#### Contracts Data  date type format

In [116]:
# list of date_cols 
contract_data_date_cols = ["Contract_Start_Date", "Expected_Contract_End_Date", "Actual_Contract_End_Date"]

# Apply the function to fix the date column
for col in contract_data_date_cols:
    contract_data[col] = contract_data[col].apply(fix_date_format)

# Fix Leap year
fix_leap_year_dates(contract_data, contract_data_date_cols)

# Chande to datetime datatype
change_to_date(contract_data, contract_data_date_cols)

# Confirm     
# app_data['Application_Date'].dtype
contract_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7598 entries, 0 to 7597
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Contract_ID                 7598 non-null   object        
 1   Application_ID              7598 non-null   object        
 2   Key                         7598 non-null   int64         
 3   Loan_Type                   7598 non-null   object        
 4   Contract_Start_Date         7598 non-null   datetime64[ns]
 5   Expected_Contract_End_Date  7598 non-null   datetime64[ns]
 6   Actual_Contract_End_Date    7598 non-null   datetime64[ns]
 7   Loan_Term                   7598 non-null   object        
 8   Total_Loan_Amount           7598 non-null   float64       
dtypes: datetime64[ns](3), float64(1), int64(1), object(4)
memory usage: 593.6+ KB


#### Contracts Snapshot date format

In [117]:
# Get date columns
contractSnapshot_date_cols = ["Snapshot Date"]

# Apply the function to fix the date column
for col in contractSnapshot_date_cols:
    contractsSnapshot_data[col] = contractsSnapshot_data[col].apply(fix_date_format)

# Fix Leap year
fix_leap_year_dates(contractsSnapshot_data, contractSnapshot_date_cols)

# Change to datetime datatype
change_to_date(contractsSnapshot_data, contractSnapshot_date_cols)

#### CRB Data date format

In [118]:
# Get date columns
crb_date_cols = ["Application_Date"]

# Apply the function to fix the date column
for col in crb_date_cols:
    crb_data[col] = crb_data[col].apply(fix_date_format)

# Fix Leap year
fix_leap_year_dates(crb_data, crb_date_cols)

# Change to datetime datatype
change_to_date(crb_data, crb_date_cols)

#### Current and Savings date format

In [ ]:
# curr_savings_data
# Apply the function to fix the date column
for col in curr_savings_date_cols:
    curr_savings_data[col] = curr_savings_data[col].apply(fix_date_format)

# Fix Leap year
fix_leap_year_dates(curr_savings_data, curr_savings_date_cols)
                    

# Change to datetime datatype
change_to_date(curr_savings_data, curr_savings_date_cols)

# Confirm 
curr_savings_data.info()

In [ ]:
curr_savings_data.isnull().sum()

ID                                        0
Account_Number                            0
Application ID                            0
Month (YYYY-MM)                           0
Open Date                                 0
Average Balance                           0
Overdraft Limit                           0
Number of Days in Debit during Month      0
Number of Days in Credit during Month     0
Maximum End of Day Balance                0
Minimum End of Day Balance                0
Sum of Monthly Debit Transactions         0
Sum of Monthly Credit Transactions        0
Last Transaction Date                     0
Number of ATM Deposits                    0
Number of ATM Withdrawals                 0
Number of Cheque Debit Transactions       0
Number of Bounced Cheques                 0
Number of Debit Card POS Transactions     0
Number of Over the Counter withdrawals    0
Account Type                              0
Account Status                            0
Number of Joint Holders         

### Data Merging to form 1 dataset

In [ ]:
contract_data.columns

Index(['Contract_ID', 'Application_ID', 'Key', 'Loan_Type',
       'Contract_Start_Date', 'Expected_Contract_End_Date',
       'Actual_Contract_End_Date', 'Loan_Term', 'Total_Loan_Amount'],
      dtype='object')

In [ ]:
# Merge on Contract_ID (ContractsData) and Contract Id (ContractsSnapshotData)
contract_merged_df = contract_data.merge(contractsSnapshot_data, left_on="Contract_ID", right_on="Contract_ID", how="inner")

# Drop duplicate column after merging
contract_merged_df.drop(columns=["Contract_ID"], inplace=True)

# Display the merged DataFrame
contract_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 121952 entries, 0 to 121951
Data columns (total 13 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Application_ID              121952 non-null  object        
 1   Key                         121952 non-null  int64         
 2   Loan_Type                   121952 non-null  object        
 3   Contract_Start_Date         121952 non-null  datetime64[ns]
 4   Expected_Contract_End_Date  121952 non-null  datetime64[ns]
 5   Actual_Contract_End_Date    121952 non-null  datetime64[ns]
 6   Loan_Term                   121952 non-null  object        
 7   Total_Loan_Amount           121952 non-null  float64       
 8   Snapshot Date               121952 non-null  datetime64[ns]
 9   Outstanding Amount          121952 non-null  float64       
 10  Due Amount                  121952 non-null  float64       
 11  Number of Days Past Due     121952 non-

In [ ]:
# Merge on Application_ID (merged df) and Application_ID (Application Data)
app_merged_df = contract_merged_df.merge(app_data, left_on="Application_ID", right_on="Application_ID", how="inner")

# Display the merged DataFrame summary
app_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 121952 entries, 0 to 121951
Data columns (total 27 columns):
 #   Column                                   Non-Null Count   Dtype         
---  ------                                   --------------   -----         
 0   Application_ID                           121952 non-null  object        
 1   Key                                      121952 non-null  int64         
 2   Loan_Type                                121952 non-null  object        
 3   Contract_Start_Date                      121952 non-null  datetime64[ns]
 4   Expected_Contract_End_Date               121952 non-null  datetime64[ns]
 5   Actual_Contract_End_Date                 121952 non-null  datetime64[ns]
 6   Loan_Term                                121952 non-null  object        
 7   Total_Loan_Amount                        121952 non-null  float64       
 8   Snapshot Date                            121952 non-null  datetime64[ns]
 9   Outstanding Amount        

In [ ]:
# Merge on Application_ID (app merged df) and crb Application_ID (Application Data)
crb_merged_df = app_merged_df.merge(crb_data, left_on="Application_ID", right_on="Application_ID", how="inner")

# Display the merged DataFrame summary
crb_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 91853 entries, 0 to 91852
Data columns (total 30 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Application_ID                           91853 non-null  object        
 1   Key                                      91853 non-null  int64         
 2   Loan_Type                                91853 non-null  object        
 3   Contract_Start_Date                      91853 non-null  datetime64[ns]
 4   Expected_Contract_End_Date               91853 non-null  datetime64[ns]
 5   Actual_Contract_End_Date                 91853 non-null  datetime64[ns]
 6   Loan_Term                                91853 non-null  object        
 7   Total_Loan_Amount                        91853 non-null  float64       
 8   Snapshot Date                            91853 non-null  datetime64[ns]
 9   Outstanding Amount                     

In [ ]:
# Merge on Application_ID (crb merged df) and curr savings Application_ID (Application Data)
# curr_savings_merged_df = crb_merged_df.merge(curr_savings_data, left_on="Application_ID", right_on="Application ID", how="inner")

# Drop duplicate column after merging
# crb_curr_savings_merged_dfmerged_df.drop(columns=["Application_ID"], inplace=True)

# Display the merged DataFrame summary
# curr_savings_merged_df.info()

### Basic Overview - data EDA

In [ ]:
# Check dataset shape (df.shape)


In [ ]:
# Check data types (df.dtypes)


In [ ]:
#Identify unique values in categorical columns (df.nunique())


### Summary statistics

In [ ]:
# Numerical Data: Check for mean, median, min, max, and standard deviation .describe()

In [ ]:
# Numerical: Check skewness df.skew()

In [ ]:
# Categorical data: Frequency distribution (df["colomn"].value_counts())

# Method 2

## Selecting relevant columns/ dropping irrelevant columns
The relevant columns for loan default prediction include:

Relevant Columns:
* ApplicationData.csv: Loan_Approval_Status, Monthly_Income, Number_of_Dependents, Time_at_Current_Employment_(Months), Employment_Status
* ContractsData.csv: Loan_Type, Loan_Term, Total_Loan_Amount
* ContractsSnapshotData.csv: Outstanding Amount, Number of Days Past Due, Contract Status
* CRB Data.csv: CRB Score, CRB Grade
* Current and Savings Account Data.csv: Average Balance, Overdraft Limit, Sum of Monthly Credit Transactions, Number of Bounced Cheques


In [ ]:
# # Select Relevant columns
# new_app_data= app_data[["Application_ID", "Loan_Approval_Status", "Monthly_Income", 
#                         "Number_of_Dependents", "Time_at_Current_Employment_(Months)", 
#                         "Employment_Status", "Government_Employee"]]

# new_contract_data = contract_data[["Contract_ID", "Application_ID", 
#                                    "Contract_Start_Date", 
#                                    "Expected_Contract_End_Date", "Loan_Type", 
#                                    "Loan_Term", "Total_Loan_Amount"]]

# new_ContractsSnapshot_data = contractsSnapshot_data[["Contract_ID", 
#                                                      "Outstanding Amount", "Due Amount",
#                                                      "Number of Days Past Due", 
#                                                      "Contract Status"]]

# new_crb_data = crb_data[["Application_ID", "CRB Score", "CRB Grade"]]

# new_curr_savings_data = curr_savings_data[["Application ID", "Average Balance", 
#                                            "Overdraft Limit", 
#                                            "Number of Days in Debit during Month",
#                                            "Number of Days in Credit during Month",
#                                            "Sum of Monthly Debit Transactions",
#                                            "Sum of Monthly Credit Transactions",
#                                            "Sum of Bounced Cheque transactions", 
#                                            "Number of Bounced Cheques",
#                                            "Sum of Over the Counter Withdrawals",
#                                            "Account Type"]]

KeyError: "['Sum of Over the Counter Withdrawals', 'Sum of Bounced Cheque transactions'] not in index"

## Check duplicates

In [ ]:
# check duplicates on the datasets
# app_data.duplicated().sum()
new_app_data.duplicated().sum()

0

In [ ]:
# contract_data.duplicated().sum()
new_contract_data.duplicated().sum()

0

In [ ]:
# contractsSnapshot_data.duplicated().sum() # 84
# new_ContractsSnapshot_data.duplicated().sum() # 15724 ??

15724

In [ ]:
# crb_data.duplicated().sum() # 0
# new_crb_data.duplicated().sum() # 75??

75

In [ ]:
# # curr_savings_data.duplicated().sum() # 0
# new_curr_savings_data.duplicated().sum() # 194,048??

194048

## Feature transformation - log transformation/OHE/etc.

## Feature selection - (PCA/LASSO)

# Exploration Data Analysis

Univariate Analysis

Histograms & KDE Plots: Distribution of numerical features
Box Plots: Detect outliers in features like Monthly Income, Loan Amount
Bar Charts: Visualize frequency distribution for categorical variables like Loan Type, Contract Status, CRB Grade

Bivariate & Multivariate Analysis
* Correlation Matrix: Find relationships between features
* Use df.corr() & visualize with heatmaps (seaborn.heatmap())
* Scatter Plots: Identify linear/non-linear relationships
Pair Plots: Detect hidden patterns between numeric features
GroupBy Analysis: Compare default rates across different CRB grades
